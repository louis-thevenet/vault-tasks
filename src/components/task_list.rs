use color_eyre::Result;
use ratatui::{prelude::*, widgets::*};
use tokio::sync::mpsc::UnboundedSender;
use tracing::info;

use super::Component;
use crate::{action::Action, config::Config};

#[derive(Default)]
pub struct TaskList {
    command_tx: Option<UnboundedSender<Action>>,
    config: Config,
}
impl TaskList {
    pub fn new() -> Self {
        Self::default()
    }
}
impl Component for TaskList {
    fn draw(&mut self, frame: &mut Frame, area: Rect) -> Result<()> {
        let surrounding_block = Block::default()
            .borders(Borders::ALL)
            .title("Here is a list of tasks");
        let items = ["Task 1", "Task 2", "Task 3"];
        let list = List::new(items).block(surrounding_block);
        frame.render_widget(list, area);
        Ok(())
    }
}
