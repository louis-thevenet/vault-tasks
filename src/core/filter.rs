use crate::core::TasksConfig;

use super::{
    date::Date,
    parser::task::parse_task,
    task::{State, Task},
    vault_data::VaultData,
};

#[derive(Default, PartialEq, Eq, Debug)]
pub struct Filter {
    pub task: Task,
    state: Option<State>,
}

/// Parses a [`Task`] from an input `&str`. Returns the `Task` and whether the input specify a task state (- [X] or - [ ]) or not.
#[must_use]
pub fn parse_search_input(input: &str, config: &TasksConfig) -> Filter {
    // Are searching for a specific state ?
    let has_state = input.starts_with("- [");

    // Make the input parsable, add a task state if needed
    let input_value = format!("{}{}", if has_state { "" } else { "- [ ]" }, input);

    // Parse the input
    let task = match parse_task(&mut input_value.as_str(), String::new(), config) {
        Ok(t) => t,
        Err(_e) => Task {
            name: String::from("Incomplete search prompt"),
            ..Default::default()
        },
    };
    Filter {
        task: task.clone(),
        state: if has_state { Some(task.state) } else { None },
    }
}

fn filter_task(task: &Task, filter: &Filter) -> bool {
    let state_match = filter.state.is_none()
        || filter.state.clone().is_some_and(|state| {
            // This is not really satisfying as you can't
            // match only incomplete tasks for example
            // I should add a config option for it
            matches!(
                (state, &task.state),
                (
                    State::ToDo | State::Incomplete,
                    State::ToDo | State::Incomplete
                ) | (State::Done | State::Canceled, State::Done | State::Canceled)
            )
        });

    let name_match = names_match(&task.name, &filter.task.name);

    let today_flag_match = if filter.task.is_today {
        task.is_today
    } else {
        true
    };

    let date_match = match (task.due_date.clone(), filter.task.due_date.clone()) {
        (_, None) => true,
        (Some(Date::DayTime(task_date)), Some(Date::DayTime(search_date)))
            if task_date == search_date =>
        {
            true
        }
        (Some(Date::Day(task_date)), Some(Date::Day(search_date))) if task_date == search_date => {
            true
        }
        (_, _) => false,
    };

    let tags_match = filter
        .task
        .tags
        .clone()
        .unwrap_or_default()
        .iter()
        .all(|t| {
            task.tags
                .clone()
                .unwrap_or_default()
                .iter()
                .any(|x| x.to_lowercase().contains(&t.to_lowercase()))
        });

    let priority_match = if filter.task.priority > 0 {
        filter.task.priority == task.priority
    } else {
        true
    };

    state_match && name_match && today_flag_match && date_match && tags_match && priority_match
}

fn names_match(name: &str, filter_name: &str) -> bool {
    if filter_name.is_empty() {
        true
    } else {
        // for each word of the filter_task, if at least one
        // matches in the task, then validate
        filter_name
            .to_lowercase()
            .split_whitespace()
            .filter(|w| name.to_lowercase().contains(w))
            .count()
            > 0
    }
}

/// Collects all tasks matching the provided `Filter` from the `VaultData` in a `Vec<Task>`.
/// If `explore_children` is true, it will also explore subtasks of tasks.
fn filter_to_vec_layer(
    vault_data: &VaultData,
    task_filter: &Filter,
    explore_children: bool,
    res: &mut Vec<Task>,
) {
    match vault_data {
        VaultData::Directory(_, children) | VaultData::Header(_, _, children) => {
            for c in children {
                filter_to_vec_layer(&c.clone(), task_filter, explore_children, res);
            }
        }
        VaultData::Task(task) => {
            if explore_children {
                task.subtasks.iter().for_each(|t| {
                    filter_to_vec_layer(
                        &VaultData::Task(t.clone()),
                        task_filter,
                        explore_children,
                        res,
                    );
                });
            }

            if filter_task(task, task_filter) {
                res.push(task.clone());
            }
        }
        VaultData::Tracker(_tracker) => (), // Don't collect trackers in the result
                                            // It's only used by the Filter and Calendar
                                            // tabs and we don't want to display trackers there
    }
}

/// Will return a `Vec<Task>` matching the given `Filter` from the `VaultData`
pub fn filter_to_vec(vault_data: &VaultData, filter: &Filter) -> Vec<Task> {
    let res = &mut vec![];
    filter_to_vec_layer(vault_data, filter, true, res);
    res.clone()
}

/// Filters a `VaultData` structure based on the provided `Filter`.
/// Only keeps the `VaultData` entries that match the filter criteria.
pub fn filter(vault_data: &VaultData, task_filter: &Filter) -> Option<VaultData> {
    match vault_data {
        VaultData::Header(level, name, children) => {
            let mut actual_children = vec![];
            for child in children {
                let child_clone = child.clone();
                if let Some(child) = filter(&child_clone, task_filter) {
                    actual_children.push(child);
                }
            }
            if actual_children.is_empty() {
                None
            } else {
                Some(VaultData::Header(*level, name.to_string(), actual_children))
            }
        }
        VaultData::Directory(name, children) => {
            let mut actual_children = vec![];
            for child in children {
                let child_clone = child.clone();
                if let Some(child) = filter(&child_clone, task_filter) {
                    actual_children.push(child);
                }
            }
            if actual_children.is_empty() {
                None
            } else {
                Some(VaultData::Directory(name.to_string(), actual_children))
            }
        }
        VaultData::Task(task) => {
            if filter_task(task, task_filter) {
                Some(vault_data.clone())
            } else {
                let mut actual_children = vec![];
                for child in &task.subtasks {
                    if let Some(VaultData::Task(child)) =
                        filter(&VaultData::Task(child.clone()), task_filter)
                    {
                        actual_children.push(child);
                    }
                }
                if actual_children.is_empty() {
                    return None;
                }
                Some(VaultData::Task(Task {
                    subtasks: actual_children,
                    ..task.clone()
                }))
            }
        }
        VaultData::Tracker(tracker) => {
            // We keep the tracker if its name matches the filter task's name
            // But we don't look at the task's state
            // I might want to refactor the Filter to allow parsing a Tracker from
            // the input string later.
            if names_match(&tracker.name, &task_filter.task.name) {
                Some(VaultData::Tracker(tracker.clone()))
            } else {
                None
            }
        }
    }
}

#[cfg(test)]
mod tests {
    use chrono::NaiveDate;

    use crate::core::{
        TasksConfig,
        date::Date,
        filter::{Filter, filter},
        task::{State, Task},
        vault_data::VaultData,
    };

    use super::{filter_to_vec, parse_search_input};

    #[test]
    fn parse_search_input_test() {
        let input = "- [ ] #tag today name p5";
        let config = TasksConfig::default();
        let res = parse_search_input(input, &config);
        let expected = Filter {
            task: Task {
                due_date: Some(Date::Day(chrono::Local::now().date_naive())),
                name: String::from("name"),
                priority: 5,
                state: State::ToDo,
                tags: Some(vec![String::from("tag")]),
                ..Default::default()
            },
            state: Some(State::ToDo),
        };
        assert_eq!(expected, res);
    }

    #[test]
    fn parse_search_input_no_state_test() {
        let input = "#tag today name p5";
        let config = TasksConfig::default();
        let res = parse_search_input(input, &config);
        let expected = Filter {
            task: Task {
                due_date: Some(Date::Day(chrono::Local::now().date_naive())),
                name: String::from("name"),
                priority: 5,
                state: State::ToDo,
                tags: Some(vec![String::from("tag")]),
                ..Default::default()
            },
            state: None,
        };
        assert_eq!(expected, res);
    }

    #[test]
    fn filter_tags_test() {
        let input = VaultData::Directory(
            "test".to_owned(),
            vec![
                VaultData::Header(
                    0,
                    "Test".to_string(),
                    vec![
                        VaultData::Header(
                            1,
                            "1".to_string(),
                            vec![VaultData::Header(
                                2,
                                "2".to_string(),
                                vec![VaultData::Task(Task {
                                    name: "test 1".to_string(),
                                    line_number: Some(8),
                                    description: Some("test\ndesc".to_string()),
                                    ..Default::default()
                                })],
                            )],
                        ),
                        VaultData::Header(
                            1,
                            "1.2".to_string(),
                            vec![
                                VaultData::Header(3, "3".to_string(), vec![]),
                                VaultData::Header(
                                    2,
                                    "4".to_string(),
                                    vec![VaultData::Task(Task {
                                        name: "test 2".to_string(),
                                        line_number: Some(8),
                                        tags: Some(vec!["test".to_string()]),
                                        description: Some("test\ndesc".to_string()),
                                        ..Default::default()
                                    })],
                                ),
                            ],
                        ),
                    ],
                ),
                VaultData::Task(Task {
                    name: "test 3".to_string(),
                    line_number: Some(8),
                    tags: Some(vec!["test".to_string()]),
                    description: Some("test\ndesc".to_string()),
                    ..Default::default()
                }),
            ],
        );
        let expected = vec![
            Task {
                name: "test 2".to_string(),
                line_number: Some(8),
                tags: Some(vec!["test".to_string()]),
                description: Some("test\ndesc".to_string()),
                ..Default::default()
            },
            Task {
                name: "test 3".to_string(),
                line_number: Some(8),
                tags: Some(vec!["test".to_string()]),
                description: Some("test\ndesc".to_string()),
                ..Default::default()
            },
        ];
        let res = filter_to_vec(
            &input,
            &Filter {
                task: Task {
                    name: String::new(),
                    tags: Some(vec!["test".to_string()]),
                    ..Default::default()
                },
                state: None,
            },
        );
        assert_eq!(res, expected);
    }
    #[test]
    fn filter_names_test() {
        let input = VaultData::Directory(
            "test".to_owned(),
            vec![
                VaultData::Header(
                    0,
                    "Test".to_string(),
                    vec![
                        VaultData::Header(
                            1,
                            "1".to_string(),
                            vec![VaultData::Header(
                                2,
                                "2".to_string(),
                                vec![VaultData::Task(Task {
                                    name: "hfdgqskhjfg1".to_string(),
                                    line_number: Some(8),
                                    description: Some("test\ndesc".to_string()),
                                    ..Default::default()
                                })],
                            )],
                        ),
                        VaultData::Header(
                            1,
                            "1.2".to_string(),
                            vec![
                                VaultData::Header(3, "3".to_string(), vec![]),
                                VaultData::Header(
                                    2,
                                    "4".to_string(),
                                    vec![VaultData::Task(Task {
                                        name: "test 2".to_string(),
                                        line_number: Some(8),
                                        tags: Some(vec!["test".to_string()]),
                                        description: Some("test\ndesc".to_string()),
                                        ..Default::default()
                                    })],
                                ),
                            ],
                        ),
                    ],
                ),
                VaultData::Task(Task {
                    name: "test 3".to_string(),
                    line_number: Some(8),
                    tags: Some(vec!["test".to_string()]),
                    description: Some("test\ndesc".to_string()),
                    ..Default::default()
                }),
            ],
        );
        let expected = vec![
            Task {
                name: "test 2".to_string(),
                line_number: Some(8),
                tags: Some(vec!["test".to_string()]),
                description: Some("test\ndesc".to_string()),
                ..Default::default()
            },
            Task {
                name: "test 3".to_string(),
                line_number: Some(8),
                tags: Some(vec!["test".to_string()]),
                description: Some("test\ndesc".to_string()),
                ..Default::default()
            },
        ];
        let res = filter_to_vec(
            &input,
            &Filter {
                task: Task {
                    name: String::from("test"),
                    ..Default::default()
                },
                state: None,
            },
        );
        assert_eq!(res, expected);
    }
    #[test]
    fn filter_due_date_test() {
        let input = VaultData::Directory(
            "test".to_owned(),
            vec![
                VaultData::Header(
                    0,
                    "Test".to_string(),
                    vec![
                        VaultData::Header(
                            1,
                            "1".to_string(),
                            vec![VaultData::Header(
                                2,
                                "2".to_string(),
                                vec![VaultData::Task(Task {
                                    name: "hfdgqskhjfg1".to_string(),
                                    line_number: Some(8),
                                    due_date: Some(Date::Day(
                                        NaiveDate::from_ymd_opt(2020, 2, 2).unwrap(),
                                    )),
                                    description: Some("test\ndesc".to_string()),
                                    ..Default::default()
                                })],
                            )],
                        ),
                        VaultData::Header(
                            1,
                            "1.2".to_string(),
                            vec![
                                VaultData::Header(3, "3".to_string(), vec![]),
                                VaultData::Header(
                                    2,
                                    "4".to_string(),
                                    vec![VaultData::Task(Task {
                                        name: "test 2".to_string(),
                                        line_number: Some(8),
                                        tags: Some(vec!["test".to_string()]),
                                        description: Some("test\ndesc".to_string()),
                                        ..Default::default()
                                    })],
                                ),
                            ],
                        ),
                    ],
                ),
                VaultData::Task(Task {
                    name: "test 3".to_string(),
                    line_number: Some(8),
                    tags: Some(vec!["test".to_string()]),
                    description: Some("test\ndesc".to_string()),
                    ..Default::default()
                }),
            ],
        );
        let expected = vec![Task {
            name: "hfdgqskhjfg1".to_string(),
            line_number: Some(8),
            due_date: Some(Date::Day(NaiveDate::from_ymd_opt(2020, 2, 2).unwrap())),
            description: Some("test\ndesc".to_string()),
            ..Default::default()
        }];
        let res = filter_to_vec(
            &input,
            &Filter {
                task: Task {
                    due_date: Some(Date::Day(NaiveDate::from_ymd_opt(2020, 2, 2).unwrap())),
                    ..Default::default()
                },
                state: None,
            },
        );
        assert_eq!(res, expected);
    }
    #[test]
    fn filter_full_test() {
        let input = VaultData::Directory(
            "test".to_owned(),
            vec![
                VaultData::Header(
                    0,
                    "Test".to_string(),
                    vec![
                        VaultData::Header(
                            1,
                            "1".to_string(),
                            vec![VaultData::Header(
                                2,
                                "2".to_string(),
                                vec![VaultData::Task(Task {
                                    name: "real target".to_string(),
                                    line_number: Some(8),
                                    tags: Some(vec!["test".to_string()]),
                                    due_date: Some(Date::Day(
                                        NaiveDate::from_ymd_opt(2020, 2, 2).unwrap(),
                                    )),
                                    description: Some("test\ndesc".to_string()),
                                    ..Default::default()
                                })],
                            )],
                        ),
                        VaultData::Header(
                            1,
                            "1.2".to_string(),
                            vec![
                                VaultData::Header(3, "3".to_string(), vec![]),
                                VaultData::Header(
                                    2,
                                    "4".to_string(),
                                    vec![VaultData::Task(Task {
                                        name: "false target 2".to_string(),
                                        line_number: Some(8),
                                        tags: Some(vec!["test".to_string()]),
                                        description: Some("test\ndesc".to_string()),
                                        ..Default::default()
                                    })],
                                ),
                            ],
                        ),
                    ],
                ),
                VaultData::Task(Task {
                    name: "test 3".to_string(),
                    line_number: Some(8),
                    tags: Some(vec!["test".to_string()]),
                    description: Some("test\ndesc".to_string()),
                    ..Default::default()
                }),
            ],
        );
        let expected = vec![Task {
            name: "real target".to_string(),
            line_number: Some(8),
            due_date: Some(Date::Day(NaiveDate::from_ymd_opt(2020, 2, 2).unwrap())),
            tags: Some(vec!["test".to_string()]),
            description: Some("test\ndesc".to_string()),
            ..Default::default()
        }];
        let res = filter_to_vec(
            &input,
            &Filter {
                task: Task {
                    name: String::from("target"),
                    tags: Some(vec!["test".to_string()]),
                    due_date: Some(Date::Day(NaiveDate::from_ymd_opt(2020, 2, 2).unwrap())),
                    ..Default::default()
                },
                state: None,
            },
        );
        assert_eq!(res, expected);
    }
    #[test]
    fn filter_subtasks_test() {
        let input = VaultData::Directory(
            "test".to_owned(),
            vec![
                VaultData::Header(
                    0,
                    "Test".to_string(),
                    vec![
                        VaultData::Header(
                            1,
                            "1".to_string(),
                            vec![VaultData::Header(
                                2,
                                "2".to_string(),
                                vec![VaultData::Task(Task {
                                    name: "task".to_string(),
                                    line_number: Some(8),
                                    tags: Some(vec!["test".to_string()]),
                                    description: Some("test\ndesc".to_string()),
                                    subtasks: vec![Task {
                                        name: "subtask".to_string(),
                                        ..Default::default()
                                    }],
                                    ..Default::default()
                                })],
                            )],
                        ),
                        VaultData::Header(
                            1,
                            "1.2".to_string(),
                            vec![
                                VaultData::Header(3, "3".to_string(), vec![]),
                                VaultData::Header(
                                    2,
                                    "4".to_string(),
                                    vec![VaultData::Task(Task {
                                        name: "false target 2".to_string(),
                                        line_number: Some(8),
                                        tags: Some(vec!["test".to_string()]),
                                        description: Some("test\ndesc".to_string()),
                                        ..Default::default()
                                    })],
                                ),
                            ],
                        ),
                    ],
                ),
                VaultData::Task(Task {
                    name: "test 3".to_string(),
                    line_number: Some(8),
                    tags: Some(vec!["test".to_string()]),
                    description: Some("test\ndesc".to_string()),
                    ..Default::default()
                }),
            ],
        );
        let expected = Some(VaultData::Directory(
            "test".to_owned(),
            vec![VaultData::Header(
                0,
                "Test".to_string(),
                vec![VaultData::Header(
                    1,
                    "1".to_string(),
                    vec![VaultData::Header(
                        2,
                        "2".to_string(),
                        vec![VaultData::Task(Task {
                            name: "task".to_string(),
                            line_number: Some(8),
                            tags: Some(vec!["test".to_string()]),
                            description: Some("test\ndesc".to_string()),
                            subtasks: vec![Task {
                                name: "subtask".to_string(),
                                ..Default::default()
                            }],
                            ..Default::default()
                        })],
                    )],
                )],
            )],
        ));
        let res = filter(
            &input,
            &Filter {
                task: Task {
                    name: String::from("subtask"),
                    ..Default::default()
                },
                state: None,
            },
        );
        assert_eq!(res, expected);
    }
}
