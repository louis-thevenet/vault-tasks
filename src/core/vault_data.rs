use std::fmt::Display;

use super::task::Task;

#[derive(Debug, Clone, PartialEq, Eq)]
pub enum VaultData {
    /// Name, Content
    Directory(String, Vec<VaultData>),
    /// Name, Content
    Header(usize, String, Vec<VaultData>),
    /// Task, Subtasks
    Task(Task),
}

impl Display for VaultData {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        fn write_indent(indent_length: usize, f: &mut std::fmt::Formatter) -> std::fmt::Result {
            (1..=indent_length).try_for_each(|_| write!(f, "\t"))?;
            Ok(())
        }
        fn write_underline_with_indent(
            text: &str,
            indent_length: usize,
            f: &mut std::fmt::Formatter,
        ) -> std::fmt::Result {
            write_indent(indent_length, f)?;
            writeln!(f, "{text}")?;
            write_indent(indent_length, f)?;
            for _i in 0..(text.len()) {
                write!(f, "â€¾")?;
            }
            writeln!(f)?;
            Ok(())
        }
        fn fmt_aux(
            file_entry: &VaultData,
            f: &mut std::fmt::Formatter,
            depth: usize,
        ) -> std::fmt::Result {
            match file_entry {
                VaultData::Header(_, header, entries) => {
                    write_underline_with_indent(&header.to_string(), depth, f)?;
                    for entry in entries {
                        fmt_aux(entry, f, depth + 1)?;
                    }
                }
                VaultData::Directory(name, entries) => {
                    write_underline_with_indent(&name.to_string(), depth, f)?;
                    for entry in entries {
                        fmt_aux(entry, f, depth + 1)?;
                    }
                }
                VaultData::Task(task) => {
                    for line in task.to_string().split('\n') {
                        write_indent(depth, f)?;
                        writeln!(f, "{line}")?;
                    }

                    for subtask in &task.subtasks {
                        for line in VaultData::Task(subtask.clone()).to_string().split('\n') {
                            write_indent(depth + 1, f)?;
                            writeln!(f, "{line}")?;
                        }
                    }
                }
            }
            Ok(())
        }
        fmt_aux(self, f, 0)
    }
}
