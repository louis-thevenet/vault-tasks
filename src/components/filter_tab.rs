use std::time::Duration;

use color_eyre::eyre::bail;
use color_eyre::owo_colors::OwoColorize;
use color_eyre::Result;
use crossterm::event::{self, Event, KeyCode, KeyEvent, KeyEventKind};
use ratatui::widgets::Paragraph;
use ratatui::{prelude::*, widgets::Block};
use tokio::sync::mpsc::UnboundedSender;
use tracing::{debug, error};

use super::Component;

use crate::task_core::filter::filter;
use crate::task_core::parser::task::parse_task;
use crate::task_core::task::Task;
use crate::task_core::vault_data::VaultData;
use crate::task_core::TaskManager;
use crate::widgets::task_list::TaskList;
use crate::{action::Action, config::Config};
use tui_input::{backend::crossterm::EventHandler, Input};

#[derive(Default)]
pub struct FilterTab {
    task_mgr: TaskManager,
    matching_entries: Vec<Task>,
    command_tx: Option<UnboundedSender<Action>>,
    config: Config,
    focused: bool,
    input: Input,
    input_mode: InputMode,
}

#[derive(Default)]
enum InputMode {
    Normal,
    #[default]
    Editing,
}
impl InputMode {
    const fn invert(&self) -> Self {
        match self {
            Self::Normal => Self::Editing,
            Self::Editing => Self::Normal,
        }
    }
}
impl FilterTab {
    pub fn new() -> Self {
        Self::default()
    }
    fn update_matching_entries(&mut self) -> Result<()> {
        let input_with_state = &format!("- [ ] {}", self.input.value());
        let mut input_str = input_with_state.as_str();
        let search = match parse_task(&mut input_str, &self.config) {
            Ok(t) => t,
            Err(e) => {
                bail!("Error: {e}");
            }
        };
        self.matching_entries = filter(&self.task_mgr.tasks, &search);
        Ok(())
    }
}
impl Component for FilterTab {
    fn register_action_handler(&mut self, tx: UnboundedSender<Action>) -> Result<()> {
        self.command_tx = Some(tx);
        Ok(())
    }

    fn register_config_handler(&mut self, config: Config) -> Result<()> {
        self.task_mgr = TaskManager::load_from_config(&config)?;
        self.config = config;
        Ok(())
    }

    fn editing_mode(&self) -> bool {
        self.focused
            && match self.input_mode {
                InputMode::Normal => false,
                InputMode::Editing => true,
            }
    }
    fn update(&mut self, action: Action) -> Result<Option<Action>> {
        if self.focused {
            match action {
                Action::FocusExplorer => self.focused = false,
                Action::FocusFilter => self.focused = true,
                Action::Enter => self.input_mode = self.input_mode.invert(),
                Action::Key(key) if matches!(self.input_mode, InputMode::Editing) => match key.code
                {
                    KeyCode::Enter => self.input_mode = self.input_mode.invert(),
                    _ => {
                        self.input.handle_event(&Event::Key(key));
                        self.update_matching_entries()?
                    }
                },
                _ => (),
            }
        } else {
            match action {
                Action::FocusExplorer => self.focused = false,
                Action::FocusFilter => self.focused = true,
                _ => (),
            }
        }
        Ok(None)
    }

    fn draw(&mut self, frame: &mut Frame, _area: Rect) -> Result<()> {
        if !self.focused {
            return Ok(());
        }
        let vertical = Layout::vertical([
            Constraint::Length(1),
            Constraint::Length(3),
            Constraint::Min(0),
            Constraint::Length(1),
        ]);
        let [_header_area, search_area, content_area, _footer_areaa] = vertical.areas(frame.area());

        let width = search_area.width.max(3) - 3; // 2 for borders, 1 for cursor
        let scroll = self.input.visual_scroll(width as usize);
        match self.input_mode {
            InputMode::Normal =>
                // Hide the cursor. `Frame` does this by default, so we don't need to do anything here
                {}

            InputMode::Editing => {
                // Make the cursor visible and ask tui-rs to put it at the specified coordinates after rendering
                frame.set_cursor_position((
                    // Put cursor past the end of the input text
                    search_area
                        .x
                        .saturating_add(((self.input.visual_cursor()).max(scroll) - scroll) as u16)
                        + 1,
                    // Move one line down, from the border to the input line
                    search_area.y + 1,
                ));
            }
        }

        let input =
            Paragraph::new(self.input.value())
                .style(Style::reset())
                .block(Block::bordered().title("Input").style(Style::new().fg(
                    match self.input_mode {
                        InputMode::Editing => Color::Rgb(255, 153, 0),
                        InputMode::Normal => Color::default(),
                    },
                )))
                .scroll((0, scroll as u16));
        frame.render_widget(input, search_area);

        let entries_list = TaskList::new(
            &self.config,
            &self
                .matching_entries
                .clone()
                .iter()
                .map(|t| VaultData::Task(t.clone()))
                .collect::<Vec<VaultData>>(),
        );
        entries_list.render(content_area, frame.buffer_mut());
        Ok(())
    }
}
