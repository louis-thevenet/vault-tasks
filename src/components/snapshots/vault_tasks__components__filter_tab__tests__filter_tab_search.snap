---
source: src/components/filter_tab.rs
expression: terminal.backend()
info: "Searching for #tobuy tag"
---
"                                                                                "
"┌──────────────────────────────────────────────────────────────────────────────┐"
"│#tobuy                                                                        │"
"└──────────────────────────────────────────────────────────────────────────────┘"
"┌Found Tags───┐┌✅ 1 cup unsalted butter───────────────────────────────────────┐" Hidden by multi-width symbols: [(17, " ")]
"│toBuy        ││📅 2024/09/29                                                  │" Hidden by multi-width symbols: [(17, " ")]
"│             ││#toBuy                                                         │"
"│             │└─────────────────────────────────────────chocolate_lava_cake.md┘"
"│             │┌✅ 8 oz bittersweet chocolate, chopped─────────────────────────┐" Hidden by multi-width symbols: [(17, " ")]
"│             ││📅 2024/09/29                                                  │" Hidden by multi-width symbols: [(17, " ")]
"│             ││#toBuy                                                         │"
"│             │└─────────────────────────────────────────chocolate_lava_cake.md┘"
"│             │┌✅ 6 large eggs────────────────────────────────────────────────┐" Hidden by multi-width symbols: [(17, " ")]
"│             ││📅 2024/09/29                                                  │" Hidden by multi-width symbols: [(17, " ")]
"│             ││#toBuy                                                         │"
"│             │└─────────────────────────────────────────chocolate_lava_cake.md┘"
"│             │┌✅ 3 egg yolks─────────────────────────────────────────────────┐" Hidden by multi-width symbols: [(17, " ")]
"│             ││📅 2024/09/29                                                  │" Hidden by multi-width symbols: [(17, " ")]
"│             ││#toBuy                                                         │"
"│             │└─────────────────────────────────────────chocolate_lava_cake.md┘"
"│             │┌✅ 1 cup sugar─────────────────────────────────────────────────┐" Hidden by multi-width symbols: [(17, " ")]
"│             ││📅 2024/09/29                                                  │" Hidden by multi-width symbols: [(17, " ")]
"│             ││#toBuy                                                         │"
"│             │└─────────────────────────────────────────chocolate_lava_cake.md┘"
"│             │┌✅ 1 tbsp vanilla extract──────────────────────────────────────┐" Hidden by multi-width symbols: [(17, " ")]
"│             ││📅 2024/09/29                                                  │" Hidden by multi-width symbols: [(17, " ")]
"│             ││#toBuy                                                         │"
"│             │└─────────────────────────────────────────chocolate_lava_cake.md┘"
"│             │┌✅ 1 cup all-purpose flour─────────────────────────────────────┐" Hidden by multi-width symbols: [(17, " ")]
"│             ││📅 2024/09/29                                                  │" Hidden by multi-width symbols: [(17, " ")]
"│             ││#toBuy                                                         │"
"│             │└─────────────────────────────────────────chocolate_lava_cake.md┘"
"│             │┌✅ Confectioners' sugar, for dusting───────────────────────────┐" Hidden by multi-width symbols: [(17, " ")]
"│             ││📅 2024/09/29                                                  │" Hidden by multi-width symbols: [(17, " ")]
"│             ││#toBuy                                                         │"
"│             │└─────────────────────────────────────────chocolate_lava_cake.md┘"
"│             │┌✅ Pine boards (2 @ 1x6, 4 @ 1x2, 4 @ 1x4)─────────────────────┐" Hidden by multi-width symbols: [(17, " ")]
"│             ││📅 2024/09/30                                                  │" Hidden by multi-width symbols: [(17, " ")]
"│             ││#toBuy                                                         │"
"│             │└───────────────────────────────────────────────diy_bookshelf.md┘"
"│             │┌✅ Wood screws (1.5 inch)──────────────────────────────────────┐" Hidden by multi-width symbols: [(17, " ")]
"│             ││📅 2024/09/30                                                  │" Hidden by multi-width symbols: [(17, " ")]
"│             ││#toBuy                                                         │"
"│             │└───────────────────────────────────────────────diy_bookshelf.md┘"
"│             │┌✅ Sandpaper (grit 120)────────────────────────────────────────┐" Hidden by multi-width symbols: [(17, " ")]
"│             ││📅 2024/09/30                                                  │" Hidden by multi-width symbols: [(17, " ")]
"│             ││#toBuy                                                         │"
"│             │└───────────────────────────────────────────────diy_bookshelf.md┘"
"│             │┌✅ Paint or stain of choice────────────────────────────────────┐" Hidden by multi-width symbols: [(17, " ")]
"│             ││📅 2024/09/30                                                  │" Hidden by multi-width symbols: [(17, " ")]
"│             ││#toBuy                                                         │"
"│             │└───────────────────────────────────────────────diy_bookshelf.md┘"
"│             │                                                                 "
"│             │                                                                 "
"│             │                                                                 "
"│             │                                                                 "
"│             │                                                                 "
"│             │                                                                 "
"│             │                                                                 "
"│             │                                                                 "
"│             │                                                                 "
"│             │                                                                 "
"│             │                                                                 "
"│             │                                                                 "
"│             │                                                                 "
"│             │                                                                 "
"│             │                                                                 "
"│             │                                                                 "
"│             │                                                                 "
"│             │                                                                 "
"│             │                                                                 "
"│             │                                                                 "
"│             │                                                                 "
"│             │                                                                 "
"│             │                                                                 "
"│             │                                                                 "
"│             │                                                                 "
"│             │                                                                 "
"│             │                                                                 "
"│             │                                                                 "
"│             │                                                                 "
"│             │                                                                 "
"│             │                                                                 "
"│             │                                                                 "
"│             │                                                                 "
"│             │                                                                 "
"│             │                                                                 "
"│             │                                                                 "
"│             │                                                                 "
"│             │                                                                 "
"│             │                                                                 "
"│             │                                                                 "
"│             │                                                                 "
"│             │                                                                 "
"│             │                                                                 "
"│             │                                                                 "
"│             │                                                                 "
"└─────────────┘                                                                 "
"                         Press Enter to stop searching                          "
"                                                                                "
